backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserServiceTest.java

```diff
@@ -0,0 +1,137 @@
+                () -> assertEquals(mockUser.getUsername(), response.get().getUsername()),
+                () -> assertEquals(mockUser.getFirstName(), response.get().getFirstName()),
+                () -> assertEquals(mockUser.getLastName(), response.get().getLastName()),
+                () -> assertEquals(mockUser.isEnabled(), response.get().isEnabled()),
+                () -> assertEquals(mockUser.isEmailVerified(), response.get().isEmailVerified()),
+                () -> assertEquals(mockUser.getAttributes().size(), response.get().getAttributes().size()),
+                () -> assertThat(attributesMappedCorrectly(mockUser.getAttributes(), response.get().getAttributes())).isTrue()
+        );
+    }
+
+    private boolean attributesMappedCorrectly(Collection<UserAttributeEntity> attributes, Map<String, List<String>> dtoAttributes) {
+        attributes.forEach(attribute -> {
+            assertTrue(dtoAttributes.containsKey(attribute.getName()));
+            List<String> values = dtoAttributes.get(attribute.getName());
+            assertTrue(values.contains(attribute.getValue()));
+        });
+        return true;
+    }
+
+    @Test
+    public void testGetUserByIdNotFound() {
+        Mockito.when(userRepository.findById("non-existing-id")).thenReturn(Optional.empty());
+        Optional<UserDTO> response = userService.getUserByID("non-existing-id");
+
+        assertThat(response.isEmpty()).isTrue();
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideParamsForGetUsers")
+    public void testUserSpecificationInvocationsInGetUsers(Optional<String> email,
+                             Optional<String> firstName,
+                             Optional<String> lastName,
+                             Optional<String> search,
+                             Optional<String> username,
+                             Optional<String> organizationId) {
+        userService.getUsers(email, firstName, lastName, search, username, organizationId);
+        int emailInvocations = email.isPresent() || search.isPresent() ? 1 : 0;
+        int firstNameInvocations = firstName.isPresent() || search.isPresent() ? 1 : 0;
+        int lastNameInvocations = lastName.isPresent() || search.isPresent() ? 1 : 0;
+        int usernameInvocations = username.isPresent() || search.isPresent() ? 1 : 0;
+
+        Mockito.verify(userSpecifications, Mockito.times(1)).notServiceAccount();
+        Mockito.verify(userSpecifications, Mockito.times(1)).fromMohApplicationsRealm();
+        Mockito.verify(userSpecifications, Mockito.times(emailInvocations)).emailLike(ArgumentMatchers.anyString());
+        Mockito.verify(userSpecifications, Mockito.times(firstNameInvocations)).firstNameLike(ArgumentMatchers.anyString());
+        Mockito.verify(userSpecifications, Mockito.times(lastNameInvocations)).lastNameLike(ArgumentMatchers.anyString());
+        Mockito.verify(userSpecifications, Mockito.times(usernameInvocations)).usernameLike(ArgumentMatchers.anyString());
```

Comment:
Blackbox tests are best. They focus on what the app does, not how. Tests that check methods break easily and are hard to follow because they don't reflect real requirements.

If a test fails because I changed how something works, but not what it does, I'll drop the test. Ideally, our main functionality tests still cover us. This keeps testing straightforward and focused.

So I'm not saying to delete these tests, just to ensure that all the functional requirements are being tested. 

This kind of test, in my opinion, is only useful during development by the developer writing the code.

Reply:
Thanks, good point. I think I wanted to somehow test the method because it's an important one, but did not find a meaningful way to test it. Now I understand what you mean. Every time I add more params to the function or change the logic a bit I need to change the test as well  - so probably will delete it.